bedrock/base/middleware.py

```diff
@@ -320,4 +322,18 @@ def process_response(self, request, response):
                         response._csp_config = config.get("DIRECTIVES", {})
                     break
 
-        return super().process_response(request, response)
+        response = super().process_response(request, response)
+
+        # Check enforced CSP reporting.
+        report_percentage = getattr(settings, "CONTENT_SECURITY_POLICY", {}).get("REPORT_PERCENTAGE", 100)
+        remove_header = random.randint(0, 99) >= report_percentage
+        if remove_header:
+            del response.headers[csp.HEADER]
```

Comment:
Maybe I'm understanding this completely wrong so please don't take it too seriously as I can be completely off… but this reads to me as "if outside of reporting sample, remove CSP header completely" — which is probably not what is the intent? The reporting sample %% should only filter out the report-to directive from the CSP header, but keep the policy rules still enforced on client.

(While that could make sense for report-only removing the whole header, I still reckon that for debug purposes even the report-only can be sent with no report-to value if that's not against the specs…)

It seems to be confirmed by tests as: https://github.com/mozilla/bedrock/blob/417ab93ebae46acd47d3817585fc46d1f8949989/bedrock/base/tests/test_middleware.py#L336-L340
checks for the whole header not being present in 0% reporting, but the header should be there, just not its `report-to` directive; this also seems to be confirmed by `check_csp_report.py` run in the OP: 
```
100 requests made
defaultdict(<class 'int'>, {'No CSP header': 50, 'report-uri present': 50})
```
compared to the previous PR:

```
100 requests made
defaultdict(<class 'int'>, {'report-uri present': 50, 'report-uri not present': 50})
```

(present vs. not preset, but no **"No CSP header"** which would indicate an issue)

Reply:
Excellent catch! Thank you. I had my head in the details too far when I worked on this and you are absolutely correct... we want to clear the `report-uri` value only, not remove the whole header. I'll fix this up soon.

Reply:
👍 Glad I'm not just seeing things;) BTW I've also used report-to vs. report-url somewhat deliberately, so take that with a pinch of salt, there's probably just the older single header endpoint used, which makes it all a bit easier (not having to wrestle another header or two for the endpoint definition) — i just never remember which is which so I use the directive names sort of interchangeably O;)

Reply:
Updated

Reply:
One last Q out of ignorance: This is actually already in [`csp/contrib/rate_limiting.py`](https://github.com/mozilla/django-csp/blob/45cb61a5e762d3f4c8ee62e414302e2978445b93/csp/contrib/rate_limiting.py) — wouldn't be better to have that only in one place, either using the contrib rate-limited middleware, or folding it back into the main code somehow? (As the conf value for it, and the docs, present it as a core feature of the package: [`configuration.rst#policy-settings`](https://github.com/mozilla/django-csp/blob/45cb61a5e762d3f4c8ee62e414302e2978445b93/docs/configuration.rst#policy-settings) ~ only when you find your way to [`reports.rst#throttling-the-number-of-reports`](https://github.com/mozilla/django-csp/blob/45cb61a5e762d3f4c8ee62e414302e2978445b93/docs/reports.rst#throttling-the-number-of-reports) that explains the contrib way.)

I understand it was historically outside of the main code for testing in the wild, but this seems as a viable feature long-term, worth keeping in…?

Reply:
That's a good call-out that we should mention the middleware when we mention `REPORT_PERCENTAGE`. I'm hesitant to add more code to the middleware that would run on every request/response if it is unneeded. And it isn't obvious that this is a feature that everyone would need.

I've opened the django-csp issue 232 to address the documentation issue.

As for subclassing the middleware, yes, but there's currently a bug I need to address and roll a new release for. But since we are not reporting yet and we won't be reporting at 100%, I can subclass the middleware to avoid duplication. But this was a good place to test the fix in the meantime. More to come here soon.

Reply:
I was kinda thinking about resolving the 100<100 issue along the way — I don't know it enough to assess if that makes any perf impact, but my idea was basically (if that was part of the main csp middleware):

- is it set to `0`? skip the rand (rate-limiting) and just always remove the report-to from directives
- is it set to say `>99` or unset/`None`? skip the rand (and rate-limiting code altogether) and break early not touching anything
- only if set to a value that needs the 0..100 rand runs then and only then go ahead and do the rate-limiting dance…

Couldn't that solve both the unnecessary random code if 0/100 clearly set (and fixing the current 100% edge case), and not running the extra code if not configured — thus not needing to keep the feature separately in contrib?